File: server/NexusGit/RobloxInstance/Partitions.cs
/*
 * TheNexusAvenger
 *
 * Stores a map of Roblox instances with associated keys.
 */

using System.Collections.Generic;
using Newtonsoft.Json;

namespace NexusGit.RobloxInstance
{
    /*
     * Class representing a build for Roblox instances.
     */
    public class PartitionsBuilder
    {
        /*
         * Returns an ISerializable object from a string.
         */
        public Partitions Deserialize(string data)
        {
            return JsonConvert.DeserializeObject<Partitions>(data);
        }
    }

    /*
     * Class representing a set of partitions.
     */
    public class Partitions
    {
        public string Type = "Partitions";
        public Dictionary<string,Instance> Instances;
        
        /*
         * Creates a partitions object.
         */
        public Partitions()
        {
            this.Instances = new Dictionary<string,Instance>();
        }

        /*
         * Adds an instance.
         */
        public void AddInstance(string name,Instance instance)
        {
            this.Instances.Add(name,instance);
        }

        /*
         * Returns the instance for the game.
         */
        public Instance GetInstance(string name)
        {
            // Return the instance if it exists.
            if (this.Instances.ContainsKey(name))
            {
                return this.Instances[name];
            }

            // Return null (not found).
            return null;
        }

        /*
         * Returns the ISerializable as a string.
         */
        public string Serialize()
        {
            return JsonConvert.SerializeObject(this,Formatting.Indented);
        }
    }
}

File: client/src/Serialization/Partitions.lua
--[[
TheNexusAvenger

Stores partitions of instances.
--]]

local NexusGit = require(script.Parent.Parent):GetContext(script)
local NexusInstance = NexusGit:GetResource("NexusPluginFramework.NexusInstance.NexusInstance")
local RobloxAPI = NexusGit:GetResource("Serialization.RobloxAPI")
local InstanceSerializier = NexusGit:GetResource("Serialization.InstanceSerializier")

local HttpService = game:GetService("HttpService")

local Partitions = NexusInstance:Extend()
Partitions:SetClassName("Partitions")



--[[
Creates a partitions object.
--]]
function Partitions:__new()
	self.Partitions = {}
end

--[[
Returns the data for a partition.
--]]
function Partitions:GetData(Name)
	return self.Partitions[Name]
end

--[[
Returns the instance for a partition.
--]]
function Partitions:GetInstance(Name)
	return InstanceSerializier:Deserialize(self.Partitions[Name])
end

--[[
Stores data in a partition.
--]]
function Partitions:StoreData(Name,Data)
	self.Partitions[Name] = Data
end

--[[
Stores an Instance in a partition.
--]]
function Partitions:StoreInstance(Name,Ins)
	self:StoreData(Name,InstanceSerializier:Serialize(Ins))
end

--[[
Serializes the partitions to a JSON string.
--]]
function Partitions:ToJSON()
	return HttpService:JSONEncode({
		Type = "Partitions",
		Instances = self.Partitions,
	})
end

--[[
Deserailizes the partitions from a JSON string.
--]]
function Partitions:FromJSON(JSONString)
	--Deserialize the string.
	local PartitionsData = HttpService:JSONDecode(JSONString)
	
	--Add the data.
	local NewPartitions = Partitions.new()
	for Name,PartitionData in pairs(PartitionsData.Instances) do
		NewPartitions:StoreData(Name,PartitionData)
	end
	
	--Return the new partitions.
	return NewPartitions
end



return Partitions